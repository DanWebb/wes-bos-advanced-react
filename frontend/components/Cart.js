import React from 'react';
import {Query, Mutation} from 'react-apollo';
import gql from 'graphql-tag';
import CartStyles from './styles/CartStyles';
import Supreme from './styles/Supreme';
import CloseButton from './styles/CloseButton';
import SickButton from './styles/SickButton';

const LOCAL_STATE_QUERY = gql`
	query LOCAL_STATE_QUERY {
		cartOpen @client
	}
`;

const TOGGLE_CART_MUTATION = gql`
	mutation TOGGLE_CART_MUTATION {
		toggleCart @client
	}
`;

const Cart = () => (
	<Mutation mutation={TOGGLE_CART_MUTATION}>
		{(toggleCart) => (
			<Query query={LOCAL_STATE_QUERY}>
				{({data}) => (
					<CartStyles open={data.cartOpen}>
						<header>
							<CloseButton title="close" onClick={toggleCart}>
								&times;
							</CloseButton>
							<Supreme>Your Cart</Supreme>
							<p>You have _ items in your cart</p>
						</header>
						<footer>
							<p>$10.00</p>
							<SickButton>Checkout</SickButton>
						</footer>
					</CartStyles>
				)}
			</Query>
		)}
	</Mutation>
);

export {LOCAL_STATE_QUERY, TOGGLE_CART_MUTATION};
export default Cart;
