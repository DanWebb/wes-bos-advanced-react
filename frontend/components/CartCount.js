import React from 'react';
import styled from 'styled-components';
import {TransitionGroup, CSSTransition} from 'react-transition-group';

const AnimationStyles = styled.span`
	position: relative;

	.count {
		display: block;
		position: relative;
		transition: all 0.5s;
		backface-visibility: hidden;
	}

	.count-enter {
		transform: rotateX(0.5turn);
	}

	.count-enter-active {
		transform: rotateX(0);
	}

	.count-exit {
		top: 0;
		position: absolute;
		transform: rotateX(0);
	}

	.count-exit-active {
		transform: rotateX(0.5turn);
	}
`;

const Dot = styled.div`
	background: ${props => props.theme.red};
	color: #fff;
	border-radius: 50%;
	padding: 0.5rem;
	line-height: 2rem;
	min-width: 3rem;
	margin-left: 1rem;
	font-weight: 100;
	font-feature-settings: 'tnum';
	font-variant-numeric: tabular-nums;
`;

const CartCount = ({count}) => (
	<AnimationStyles>
		<TransitionGroup>
			<CSSTransition 
				className="count"
				classNames="count"
				key={count}
				timeout={{enter: 500, exit: 500}}
				unmountOnExit
			>
				<Dot>{count}</Dot>
			</CSSTransition>
		</TransitionGroup>
	</AnimationStyles>
);

export default CartCount;
