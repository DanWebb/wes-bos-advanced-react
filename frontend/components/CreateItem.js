import React, {Component} from 'react';
import {Mutation} from 'react-apollo';
import gql from 'graphql-tag';
import Router from 'next/router';
import formatMoney from '../lib/formatMoney';
import Form from './styles/Form';
import Error from '../components/ErrorMessage';

const CREATE_ITEM_MUTATION = gql`
	mutation CREATE_ITEM_MUTATION(
		$title: String! 
		$description: String! 
		$price: Int! 
		$image: String 
		$largeImage: String
	) {
		createItem(
			title: $title 
			description: $description 
			price: $price 
			image: $image 
			largeImage: $largeImage
		) {
			id
		}
	}
`;

class CreateItem extends Component {
	state = {
		title: '',
		description: '',
		image: '',
		largeImage: '',
		price: 0
	}

	handleChange = e => {
		const {name, type, value} = e.target;
		const val = type === 'number' ? parseFloat(value) : value;
		this.setState({[name]: val});
	}

	handleSubmit = async (e, createItem) => {
		e.preventDefault();
		const response = await createItem();
		Router.push({
			pathname: '/item',
			query: {id: response.data.createItem.id}
		});
	}

	handleUploadFile = async e => {
		const files = e.target.files;
		const data = new FormData();
		data.append('file', files[0]);
		data.append('upload_preset', 'sickfits');
		
		const response = await fetch('https://api.cloudinary.com/v1_1/danwebb/image/upload', {
			method: 'POST',
			body: data
		});
		const file = await response.json();

		this.setState({
			image: file.secure_url || '',
			largeImage: file.eager[0].secure_url
		});
	}

	render() { 
		return (
			<Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state}>
				{(createItem, {loading, error}) => (
					<Form data-test="form" onSubmit={e => this.handleSubmit(e, createItem)}>
						<Error error={error}/>
						<fieldset disabled={loading} aria-busy={loading}>
							<label htmlFor="file">
								Image
								<input
									required
									type="file"
									id="file"
									name="file"
									placeholder="Upload an image"
									onChange={this.handleUploadFile}
								/>
								{this.state.image && <img src={this.state.image} width="200" alt="Upload Preview"/>}
							</label>

							<label htmlFor="title">
								Title
								<input
									required
									type="text"
									id="title"
									name="title"
									placeholder="Title"
									value={this.state.title}
									onChange={this.handleChange}
								/>
							</label>

							<label htmlFor="price">
								Price
								<input
									required
									type="number"
									id="price"
									name="price"
									placeholder="Price"
									value={this.state.price}
									onChange={this.handleChange}
								/>
							</label>

							<label htmlFor="description">
								Description
								<textarea
									required
									id="description"
									name="description"
									placeholder="Description"
									value={this.state.description}
									onChange={this.handleChange}
								/>
							</label>

							<button type="submit">Create</button>
						</fieldset>
					</Form>
				)}
			</Mutation>
		);
	}
}

export {CREATE_ITEM_MUTATION};
export default CreateItem;