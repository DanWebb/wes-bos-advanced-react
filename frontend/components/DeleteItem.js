import React, {Component} from 'react';
import {Mutation} from 'react-apollo';
import gql from 'graphql-tag';
import {ALL_ITEMS_QUERY} from './Items';

const DELETE_ITEM_MUTATION = gql`
	mutation DELETE_ITEM_MUTATION($id: ID!) {
		deleteItem(id: $id) {
			id
		}
	}
`;

class DeleteItem extends Component {
	handleDeleteItem = (e, deleteItem) => {
		e.preventDefault();
		if (confirm('Are you sure?')) {
			return deleteItem().catch(err => {
				alert(err.message);
			});
		}
	}

	update = (cache, payload) => {
		const data = cache.readQuery({query: ALL_ITEMS_QUERY});
		data.items = data.items.filter(item => item.id !== payload.data.deleteItem.id);
		cache.writeQuery({query: ALL_ITEMS_QUERY, data});
	}

	render() { 
		return (
			<Mutation 
				mutation={DELETE_ITEM_MUTATION}
				variables={{id: this.props.id}}
				update={this.update}
			>
				{(deleteItem, {error}) => (
					<button onClick={e => this.handleDeleteItem(e, deleteItem)}>
						{this.props.children}
					</button>
				)}
			</Mutation>
		);
	}
}

export {DELETE_ITEM_MUTATION};
export default DeleteItem;