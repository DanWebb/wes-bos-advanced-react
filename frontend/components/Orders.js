import React, {Component} from 'react';
import {Query} from 'react-apollo';
import gql from 'graphql-tag';
import {formatDistance} from 'date-fns';
import Link from 'next/link';
import styled from 'styled-components';
import formatMoney from '../lib/formatMoney'
import Error from './ErrorMessage';
import OrderItemStyles from './styles/OrderItemStyles';

const ORDERS_QUERY = gql`
	query ORDERS_QUERY {
		orders {
			id
			total 
			createdAt
			user {
				id
			}
			items {
				id
				title
				price
				description
				quantity
				image
			}
		}
	}
`;

const OrderUl = styled.ul`
	display: grid;
	grid-gap: 4rem;
	grid-template-columns: repeat(auto-fit, minmax(40%, 1fr));
`;

class Orders extends Component {
	render() {
		return (
			<Query query={ORDERS_QUERY}>
				{({data: {orders}, error, loading}) => {
					if (error) return <Error error={error}/>
					if (loading) return <p>Loading...</p>
					console.log(orders);
					return (
						<div>
							<h2>You have {orders.length} orders</h2>
							<OrderUl>
								{orders.map(order => (
									<OrderItemStyles key={order.id}>
										<Link href={{pathname: '/order', query: {id: order.id}}}>
											<a>
												<div className="order-meta">
													<p>{order.items.reduce((a, b) => a + b.quantity, 0)} Items</p>
													<p>{order.items.length} products</p>
													<p>{formatDistance(order.createdAt, new Date())}</p>
													<p>{formatMoney(order.total)}</p>
												</div>
												<div className="images">
													{order.items.map(item => (
														<img key={item.id} src={item.image} alt={item.title}/>
													))}
												</div>
											</a>
										</Link>
									</OrderItemStyles>
								))}
							</OrderUl>
						</div>
					)
				}}
			</Query>
		);
	}
};

export {ORDERS_QUERY};
export default Orders;
