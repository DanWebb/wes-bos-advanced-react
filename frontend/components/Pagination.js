import React from 'react';
import PaginationStyles from './styles/PaginationStyles';
import gql from 'graphql-tag';
import {Query} from 'react-apollo';
import Head from 'next/head';
import Link from 'next/link';
import {perPage} from '../config';

const PAGINATION_QUERY = gql`
	query PAGINATION_QUERY {
		itemsConnection {
			aggregate {
				count
			}
		}
	}
`;

const Pagination = ({page}) => (
	<Query query={PAGINATION_QUERY}>
		{({data, loading, error}) => {
			if (loading) return <p>Loading...</p>;
			const count = data.itemsConnection.aggregate.count;
			const pages = Math.ceil(count / perPage);
			return (
				<PaginationStyles>
					<Head>
						<title>Emuz | Page {page} of {pages}</title>
					</Head>
					<Link prefetch href={{pathname: 'items',query: {page: page - 1}}}>
						<a className="prev" aria-disabled={page <= 1}>ðŸ‘ˆ Prev</a>
					</Link>
					<p>page {page} of {pages}</p>
					<p>{count} Items Total</p>
					<Link prefetch href={{pathname: 'items', query: {page: page + 1}}}>
						<a aria-disabled={page >= pages}>Next ðŸ‘‰</a>
					</Link>
				</PaginationStyles>
			);
		}}
	</Query>
);

export {PAGINATION_QUERY};
export default Pagination;
