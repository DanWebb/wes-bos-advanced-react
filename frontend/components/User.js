import React from 'react';
import {Query} from 'react-apollo';
import gql from 'graphql-tag';
import propTypes from 'prop-types';

const CURRENT_USER_QUERY = gql`
	query CURRENT_USER_QUERY {
		me {
			id
			email
			name
			permissions
			orders {
				id
			}
			cart {
				id
				quantity
				item {
					id
					title
					price
					image
					description
				}
			}
		}
	}
`;

const User = props => (
	<Query {...props} query={CURRENT_USER_QUERY}>
		{payload => props.children(payload)}
	</Query>
);

User.propTypes = {
	children: propTypes.func.isRequired
};

export {CURRENT_USER_QUERY};
export default User;