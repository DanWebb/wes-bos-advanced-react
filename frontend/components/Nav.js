import React from 'react';
import Link from 'next/link';
import {Mutation} from 'react-apollo';
import calcTotalQuantity from '../lib/calcTotalQuantity';
import {TOGGLE_CART_MUTATION} from './Cart';
import NavStyles from './styles/NavStyles';
import User from './User';
import Signout from './Signout';
import CartCount from './CartCount';

const Nav = () => (
	<User>
		{({data: {me}}) => (
			<NavStyles data-test="nav">
				<Link href="/items"><a>shop</a></Link>
				{me ? (
					<>
						<Link href="/sell"><a>sell</a></Link>
						<Link href="/orders"><a>orders</a></Link>
						<Link href="/account"><a>account</a></Link>
						<Signout/>
						<Mutation mutation={TOGGLE_CART_MUTATION}>
							{(toggleCart) => (
								<button onClick={toggleCart}>
									Cart
									<CartCount count={calcTotalQuantity(me.cart)}/>
								</button>
							)}
						</Mutation>
					</>
				) : (
					<Link href="/signup"><a>signup</a></Link>
				)}

			</NavStyles>
		)}
	</User>
);

export default Nav;
