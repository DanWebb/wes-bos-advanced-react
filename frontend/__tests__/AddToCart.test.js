import {mount} from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import {MockedProvider} from 'react-apollo/test-utils';
import {ApolloConsumer} from 'react-apollo';
import AddToCart, {ADD_TO_CART_MUTATION} from '../components/AddToCart';
import {CURRENT_USER_QUERY} from '../components/User';
import {fakeUser, fakeCartItem} from '../lib/testUtils';

const me = fakeUser();
const mocks = [{
	request: {query: CURRENT_USER_QUERY},
	result: {data: {me: {...me, cart: []}}} 
}, {
	request: {query: CURRENT_USER_QUERY},
	result: {data: {me: {...me, cart: [fakeCartItem()]}}} 
}, {
	request: {query: ADD_TO_CART_MUTATION, variables: {id: 'abc123'}},
	result: {data: {addToCart: {...fakeCartItem(), quantity: 1}}}
}];

describe('<AddToCart/>', () => {
	it('renders', async () => {
		const wrapper = mount(
			<MockedProvider mocks={mocks}>
				<AddToCart id="abc123"/>
			</MockedProvider>
		);
		await wait();
		wrapper.update();
		expect(toJSON(wrapper.find('button'))).toMatchSnapshot();
	});

	it('add to cart', async () => {
		let apolloClient = apolloClient;
		const wrapper = mount(
			<MockedProvider mocks={mocks}>
				<ApolloConsumer>
					{client => {
						apolloClient = client;
						return <AddToCart id="abc123"/>;
					}}
				</ApolloConsumer>
			</MockedProvider>
		);
		await wait();
		wrapper.update();
		const {data: {me}} = await apolloClient.query({query: CURRENT_USER_QUERY});
		expect(me.cart).toHaveLength(0);
		wrapper.find('button').simulate('click')
		await wait();
		const {data: {me: meTwo}} = await apolloClient.query({query: CURRENT_USER_QUERY});
		expect(meTwo.cart).toHaveLength(1);
	});

	it('loading', async () => {
		const wrapper = mount(
			<MockedProvider mocks={mocks}>
				<AddToCart id="abc123"/>
			</MockedProvider>
		);
		await wait();
		wrapper.update();
		expect(wrapper.text()).toContain('Add To Cart');
		wrapper.find('button').simulate('click');
		expect(wrapper.text()).toContain('Adding To Cart');
	})
});