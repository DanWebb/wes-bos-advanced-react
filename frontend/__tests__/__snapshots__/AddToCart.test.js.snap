// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddToCart/> renders 1`] = `
<button
  disabled={false}
  onClick={[Function]}
>
  Add
   To Cart ðŸ›’
</button>
`;
